### DISCUSSION TITLE
How do you do integration testing using Prisma? Is it integration testing ready?

### DISCUSSION THREAD

**Discussion by aronreisx** [prisma/prisma#15509] [Q&A]
I am facing a lot of frustration using Prisma 3.14 while creating integration testing for my application. The queries' slowness is making me get unpredictable results on my controllers, each time I run my test suites it can pass or fail depending on how fast DB writings can take effect. 
I started facing issues while trying to reset the database by using the NodeJS child process, but I realize it was not the most performant way to do it. Then a decided to truncate the tables, as Prisma recommends, except for the "_prisma_migration" table of course, but even that wasn't helping me in getting a clean database environment for each test suite.

So, how do you do it guys? Is Prisma really ready for testing?

**Comment by austincrim** [comment]
Hey @aronreisx 👋,

Could you provide some more details about your testing setup?
What database are you using?
Approximately how many queries are you running per test?
Have you tried increasing the timeouts in your testing framework?

**Reply by aronreisx** [reply]
Hey @austincrim , thank you for spending you time trying to help.

So, let me explain more about my test workflow:

1. I start by raising a Postgres Docker container
2. I run Prisma migrate deploy
3. Database is seeded
4. I run Jest

- At the beginning of each test suite I authenticate an administrator user
- Make one to four database insertions per test case
- Truncate the table I used on the test suite
- Disconnect (and go to the next test suite)

The problem seems to be that the truncation of the previous test suite is, sometimes, taking effect at the middle at the next test suite.
For instance, the second test case bellow, it sometimes pass or fail, usually the failure is that I get response status 200 instead of 400.

```
import request from 'supertest';

import app from '@shared/infra/http/app';
import { databaseClient } from '@shared/infra/http/database';
import { getSeededAdminUser, truncateTables } from '@utils/database';

describe('Create Category Controller', () => {
  jest.setTimeout(30000);

  let Authorization: object;

  beforeAll(async () => {
    const authentication = await request(app)
      .post('/sessions')
      .send(getSeededAdminUser());

    Authorization = { Authorization: `Bearer ${authentication.body.token}` };
  });

  afterAll(async () => {
    await truncateTables(['categories']);
    await databaseClient.$disconnect();
  });

  it('should be able to create a new category', async () => {
    const response = await request(app)
      .post('/categories')
      .send({
        name: 'Lorem ipsum dollor',
        description: 'Lorem ipsum dollor description',
      })
      .set(Authorization);

    expect(response.status).toBe(201);
    expect(response.body).toHaveProperty('id');
  });

  it('should reject the attempt to create a duplicated category', async () => {
    await request(app)
      .post('/categories')
      .send({
        name: 'Lorem category',
        description: 'Lorem category description',
      })
      .set(Authorization);

    const response = await request(app)
      .post('/categories')
      .send({
        name: 'Lorem category',
        description: 'Lorem category description',
      })
      .set(Authorization);

    expect(response.status).toBe(400);
  });
});

```

[Here you have a Gist example of the truncate method I am using](https://gist.github.com/aronreisx/2ed729df8b0db535083456a3ac4ab260)

**Reply by austincrim** [reply]
I would try adding some logs to ensure the execution order is what you expect. If you are properly `await`ing everything, you shouldn't have test overlap, if I understand the Jest parts correctly.

I would also recommend upgrading to the latest version of Prisma.

**Reply by aronreisx** [reply]
Thank you, @austincrim. I guess the best strategy to debug this is to emit logs from queries and see if everything is being handled properly.
I'll update this discussion as soon as possible.

**Reply by vladi-stevanovic** [reply]
👋 Hello @aronreisx - I wanted to follow up here to see if perhaps you have any updates for us? Let us know when / if we can help! :)

**Comment by selimb** [comment]
I just published an article on how we're doing integration testing with Prisma. I'm pretty happy with the results. Hope this helps!

https://selimb.hashnode.dev/speedy-prisma-pg-tests

As noted in the conclusion, I think a lot of that should/can be integrated directly into Prisma, especially savepoint support (that's already tracked in #12898, but that issue hasn't seen a lot of activity lately)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
